#22036의 Python 시험대비 교재
* 이 교재는 여러 수준의 독자를 위해 제작되었습니다
* 오류나 지적사항, 건의사항은 카톡이나 [여기로](https://forms.gle/cLfQGtNjQmg2hAjm7) 문의해주세요
* 이해가 가지 않는 부분이나 추가 설명(추가 예시라던가 예제라던가)이 필요하시면 마찬가지로 문의해주세요
* '~~알고리즘 정리해주세요' '풀이방법 정리해주세요' 같은 요구사항도 마찬가지로
* 피드백 적극 반영합니다
* ~~문의좀 해줘요 자습하기싫어~~

* `print()`전에 예제로 `print()`쓰냐는거는 걍 그런갑다 해줘요... 뭐 그런거까지야... 어차피 어느정도는 다 알잖아....

### Google Colab이 처음이신 분들을 위한 설명

Google Colab은 구글에서 제공하는 무료 클라우드 기반 Jupyter 노트북 환경입니다. 아래는 Google Colab의 사용 방법에 대한 간단한 설명입니다.

1. 구글 계정 로그인<br>
Google Colab에 접속하면 로그인 화면이 나타납니다. 구글 계정으로 로그인하면 Colab을 이용할 수 있습니다.

2. 노트북 생성<br>
Google Colab의 홈 화면에서 '새 노트북'을 클릭하여 새로운 노트북을 생성할 수 있습니다. 새 노트북은 Python 3 환경으로 기본 설정되어 있습니다.

3. 코드 작성<br>
노트북 내의 셀(Cell)에 코드를 작성할 수 있습니다. 셀에 코드를 입력하면 셀 좌측에 ▶모양의'실행' 버튼을 누르거나 'Shift + Enter' 키를 눌러 코드를 실행할 수 있습니다.<br>
실행 버튼: <img src="https://github.com/sleepncaffeine/my_small_python_notes/assets/101965838/becb6857-479b-4030-9f69-a83a27525ab6"></img>
4. 라이브러리 설치<br>
Google Colab은 이미 많은 라이브러리를 기본적으로 제공합니다. 그러나 사용자가 필요한 라이브러리를 설치하려면 '!'를 붙여 셀에 명령어를 입력하면 됩니다. 예를 들어, '!pip install 라이브러리 이름'과 같은 형태로 입력합니다.

5. 파일 업로드 및 다운로드<br>
Colab 환경에서 파일을 업로드하려면 화면 좌측 상단에 위치한 '파일' 메뉴에서 '업로드'를 클릭합니다. 다운로드는 셀 내에서 코드를 작성하여 가능합니다.

6. GPU 사용<br>
Google Colab은 무료로 제공되는 클라우드 기반 환경이지만, GPU를 사용하여 빠른 연산을 할 수 있습니다. '런타임' 메뉴에서 '런타임 유형 변경'을 선택하고, 하드웨어 가속기를 GPU로 설정하여 사용할 수 있습니다.

7. 노트북 공유<br>
Google Colab에서 작성한 노트북은 다른 사용자와 공유할 수 있습니다. '공유' 버튼을 클릭하여 공유 링크를 생성하고, 이 링크를 다른 사용자에게 전달하면 됩니다.

위와 같은 방법으로 Google Colab을 이용하여 머신 러닝 모델 학습, 데이터 분석 등 다양한 작업을 수행할 수 있습니다.

<big>**대충 이 교재에선 '실행 버튼(▶)을 누르면 코드가 실행되고 실행 결과가 아래에 출력되는구나!' 정도만 알고 계시면 충분합니다.**</big>

예제도 코드 실행과 같이 입력란에 입력하고 실행 버튼 누르시면 채점이 됩니다.

# Python의 자료형(데이터 유형, 데이터 타입)

프로그래밍에서 자료형은 프로그램에서 사용되는 데이터의 유형 또는 종류를 나타냅니다. 파이썬은 동적으로 유형이 지정된 언어이며, 변수를 선언할 때 데이터 유형을 지정할 필요가 없습니다. 대신, 파이썬은 변수에 할당하는 값에 기반하여 적절한 자료형형을 자동으로 지정합니다.

## 1. 수치/숫자 자료형

숫자 자료형은 파이썬에서 숫자를 나타내는 데 사용됩니다. 파이썬의 세 가지 기본 숫자 자료형은 다음과 같습니다:

* 정수 (`int`): 정수는 소수점 없이 양수와 음수의 정수입니다. 예: 2, 4, -6, 0.

* 부동 소수점 숫자 (`float`): 부동 소수점 숫자는 소수점을 가지는 실수입니다. 예: 3.14, -2.5, 0.0.

* 복소수 (`complex`): 복소수는 `a + bj` 형태로 나타내며, 여기서 `a`와 `b`는 모두 부동 소수점 숫자 또는 정수이고, `j`는 허수를 나타냅니다. 예: 2+3j, -4-2j.

## 2. 텍스트 자료형

텍스트 자료형은 문자열을 나타내는 데 사용됩니다. Python에서 텍스트 자료형은 문자를 단일 또는 이중 따옴표로 감싸는 방식으로 표현되며, 문자열(string, str)이라고 합니다. 예를 들어: "Hello, World!", 'This is a string'.

또한, `len()`함수는 문자열의 길이를 반환하는 함수입니다.<br>
<small>`len()`함수는 리스트에서도 같은 방식으로 사용할 수 있습니다.</small>

In [ ]:
print(len("Hello"))

5


## 3. 불 자료형

파이썬에서 불(부울, Boolean, Bool) 자료형은 True와 False 값을 나타내는 데 사용됩니다. 불 값은 종종 조건문에서 사용되며, 특정 조건이 참(True)인 경우에만 코드 블록이 실행됩니다.

## 예시들

In [ ]:
# 정수
a = 5
b = -3
c = 0

# 부동소수점 실수
x = 3.14
y = -2.5
z = 0.0

# 복소수
p = 2 + 3j
q = -4 - 2j

# 문자열
s1 = "Hello, World!"
s2 = 'This is a string'

# 불
is_true = True
is_false = False

# 아래 예시들이 정상적으로 돌아가려면 이 셀이 실행되어야 합니다

변수의 자료형을 판단할 때 `type()`과 같은 내장 함수를 사용할 수도 있습니다:

In [ ]:
# Checking the data type of a variable
print(type(a))
print(type(x))
print(type(s1))
print(type(is_true))

<class 'int'>
<class 'float'>
<class 'str'>
<class 'bool'>


## 문자열 갖고놀기

파이썬에서는 이러한 자료형을 조작하기 위해 다양한 내장 함수와 연산자를 사용할 수 있습니다. 예를 들어, 두 개의 문자열을 연결하는 데 `+` 연산자를 사용하거나, 문자열을 여러 번 반복하는 데 `*` 연산자를 사용할 수 있습니다. 다음은 일부 예시입니다:

In [ ]:
# Concatenation of strings
s3 = s1 + " " + s2
print(s3)

# Repetition of strings
s4 = "Hello " * 3
print(s4)

Hello, World! This is a string
Hello Hello Hello 


### 문자열의 인덱싱과 슬라이싱

파이썬에서 문자열은 문자의 순서 있는 시퀀스입니다. 이는 문자열의 각 문자가 숫자로 할당되는 것을 의미합니다. 첫 번째 문자에 대해서는 0, 두 번째 문자에 대해서는 1로 시작하여 계속해서 번호가 할당됩니다. 이 번호를 사용하여 문자열의 개별 문자에 접근할 수 있습니다.

#### 인덱싱

문자열에서 특정 문자에 접근하려면 인덱싱을 사용할 수 있습니다. 인덱싱은 대괄호 `[]`를 사용하며, 대괄호 안에 인덱스 번호를 넣습니다. 다음은 예시입니다:

In [ ]:
s = "Hello, World!"
print(s[0])

H


이 예시에서는 우리가 인덱스 `0`을 사용하여 문자열 `s`의 첫 번째 문자에 접근하고 있습니다.

문자열의 뒤에서부터 문자에 접근하려면, 마지막 문자에 대해 `-1`부터 시작하는 음수 인덱싱을 사용할 수 있습니다. 다음은 예시입니다:

In [ ]:
s = "Hello, World!"
print(s[-1])

!


이 예제에서는 인덱스 `-1`을 사용하여 문자열 s의 마지막 문자에 접근합니다.

#### 슬라이싱

슬라이싱은 문자열의 일부분을 추출하는 방법입니다. 콜론 `:`으로 구분된 인덱스 범위를 사용하여 문자열을 자를 수 있습니다. 예시를 보겠습니다:

In [ ]:
s = "Hello, World!"
print(s[0:5])

Hello


이 예제에서는 문자열 `s`를 인덱스 `0`부터 인덱스 `4`까지 슬라이싱하여 `"Hello"`라는 부분 문자열을 얻습니다. 슬라이싱에서 마지막 인덱스는 포함되지 않습니다. 따라서 인덱스 `5`의 문자는 슬라이싱된 부분 문자열에 포함되지 않습니다. (인덱스 `0`이상 `5`미만으로 이해하면 쉽습니다.)

또한, 위에서의 인덱싱과 같이 문자열의 뒤에서부터 슬라이싱하려면 음수 인덱싱을 사용할 수 있습니다. 다음은 예시입니다.

In [ ]:
s = "Hello, World!"
print(s[-6:-1])

World


이 예제에서는, 문자열 `s`를 뒤에서부터 여섯 번째 문자부터 두 번째 문자까지 자르면 `"World"`라는 부분 문자열이 생성됩니다. 마찬가지로 시작 인덱스는 포함되고 끝 인덱스는 제외됩니다.

마지막으로, 인덱스 중 하나 이상을 생략하여 문자열을 시작 또는 끝에서부터 자를 수 있습니다. 다음은 예시입니다:

In [ ]:
s = "Hello, World!"
print(s[:5])
print(s[7:])

Hello
World!


첫 예시에서는, 문자열 `s`를 처음부터 `4`번 인덱스까지 자르면서 `"Hello"` 부분 문자열을 얻습니다. 두 번째에서는, 문자열 `s`를 `7`번 인덱스부터 끝까지 자르면서 `"World!"` 부분 문자열을 얻습니다.

아, 마지막이 아니였군요.

파이썬에서 문자열을 슬라이싱할 때는 `[start:end]` 구문을 사용하여 추출할 문자의 범위를 지정할 수 있습니다. 이 구문은 슬라이스의 시작 인덱스(포함)와 끝 인덱스(제외)를 지정합니다.

`[start:end:step]` 구문을 사용하여 간격 값을 지정할 수도 있습니다. 간격 값은 말 그대로 추출될 문자 사이의 간격을 결정합니다. 예시를 살펴보겠습니다:

In [ ]:
s = "0123456789"
print(s[2:8:2])

246


이 예제에서는 인덱스 `2`에서 시작하고 인덱스 `7`에서 끝나는 (마지막 인덱스는 제외되는) 문자열 `s`를 슬라이싱하며, 간격 값을 `2`로 설정합니다. 이는 인덱스 `2`와 인덱스 `7` 사이에서 문자를 두 칸에 한 번씩 추출하므로 `"246"`이라는 부분 문자열이 반환됩니다.

간격 값을 입력하지 않으면, 기본값인 `1`로 설정됩니다.

문자열에서 문자를 역순으로 추출하기 위해 음수의 간격 값을 사용할 수 있습니다. 다음은 예시입니다:

In [ ]:
s = "0123456789"
print(s[8:2:-2])

864


이 예시에서는 문자열 `s`를 인덱스 `8`에서 시작하여 인덱스 `3`에서 끝나는 음수 간격 값이 `-2`인 슬라이싱을 수행합니다. 이는 우리가 인덱스 `8`과 인덱스 `3` 사이에서 역순으로 매 두 번째 문자를 추출하고 이를 연결하여 부분 문자열 `'864'`을 얻는 것을 의미합니다.

음수 간격 값을 사용할 때 시작과 끝 인덱스를 올바르게 지정해야만 원하는 출력을 얻을 수 있음에 유의하십시오. 예를 들어:

In [ ]:
s = "0123456789"
print(s[2:8:-2])
# 출력: ""

이 예시에서는 문자열 `s`를 인덱스 `2`에서 시작하여 인덱스 `7`에서 끝나는 음수 간격 값이 `-2`인 역순 슬라이싱을 시도합니다. 그러나 시작 인덱스가 끝 인덱스보다 크기 때문에 슬라이스는 비어 있고 출력은 빈 문자열 `""`입니다.

### 문자열 메소드

파이썬에는 문자열을 막 갖고놀 수 있는 문자열을 위한 메소드들이 있습니다. 다음은 많이 쓰이는 문자열 메소들입니다.

* `len()`: `len()`함수는 문자열의 길이를 리턴합니다.

In [ ]:
my_string = "Hello World"
print(len(my_string))  # 11

11


* `split()`: `split()`메소드는 지정된 구분자를 기준으로 문자열을 여러 개의 하위 문자열로 분할하는 데 사용됩니다(리스트로 리턴합니다). 기본적으로 구분자는 공백입니다.

In [ ]:
my_string = "Hello World"
print(my_string.split())  # ['Hello', 'World']

# 공백 말고 다른 문자 기준으로 분할하기기
my_string = "apple,banana,orange"
print(my_string.split(","))  # ['apple', 'banana', 'orange']

['Hello', 'World']
['apple', 'banana', 'orange']


* `in`: `in` 키워드는 ~~<font size=2em>메소드는 아니긴 한데 걍 여따 넣을게요</font>~~ 문자열 내에 부분 문자열이 있는지 여부를 확인하는 데 사용됩니다. 이는 불 값(True 또는 False)을 반환합니다.

In [ ]:
my_string = "Hello World"
print("World" in my_string)  # True
print("world" in my_string)  # False

True
False


* `replace()`: `replace()` 메소드는 문자열에서 하나 이상의 문자열을 새로운 문자열로 교체하는 데 사용됩니다.

In [ ]:
my_string = "Hello World"
new_string = my_string.replace("World", "Python")
print(new_string)  # Hello Python

# 두 개 이상도 한번에 교체됩니다
my_string = "Hello Hello World"
new_string = my_string.replace("Hello", "Bye")
print(new_string) # Bye Bye world

Hello Python
Bye Bye World


* `upper()`: `upper()` 메소드는 문자열의 모든 문자를 대문자로 변환하는 데 사용됩니다.

In [ ]:
my_string = "Hello World"
print(my_string.upper())  # HELLO WORLD

HELLO WORLD


* `lower()`: `lower()` 메소드는 문자열의 모든 문자를 소문자료 변환하는 데 사용됩니다.

In [ ]:
my_string = "Hello World"
print(my_string.lower())  # hello world

hello world


* `strip()`: `strip()` 메소드는 문자열의 시작과 끝에서 공백 문자를 제거하는 데 사용됩니다.

In [ ]:
my_string = "  Hello World  "
print('"' + my_string + '"')          # "  Hello World  "
print('"' + my_string.strip() + '"')  # "Hello World"

"  Hello World  "
"Hello World"


## 마무리

프로그래밍에서 자료형을 이해하는 것은 중요합니다. 이는 데이터를 효과적이고 정확하게 조작하고 사용하는 데 도움이 됩니다.

## 예제

### 예제 1

다음 코드의 실행 결과로 알맞은 것을 적으시오.
```python
name = "John"
age = 25
print("My name is {0} and I am {1} years old.".format(name, age))
```

In [ ]:
정답 = "" #@param {type:"string"}
if 정답 == "My name is John and I am 25 years old.":
    print("CORRECT")
else:
    print("WRONG")
    print("Please try again.")

### 예제 2

다음 코드의 실행 결과로 알맞은 것을 적으시오.
```python
s = "Hello, World!"
print(s[-1])
```

In [ ]:
정답 = "" #@param {type:"string"}
if 정답 == "!":
    print("CORRECT")
else:
    print("WRONG")
    print("Please try again.")

### 예제 3

다음 코드의 실행 결과로 알맞은 것을 고르시오.
```python
s = "Hello, GSA! Run away from GSA!!!"
print(s[-6:-1], s[2: 7])
```
1. `GSA!! llo, `
2. ` GSA! ello,`
3. `from , GSA!`
4. `m GSA o, GS`

In [ ]:
정답 = 0 #@param {type:"integer"}
if 정답 not in range (1,5):
    print("유효한 숫자를 입력해 주세요")
elif 정답 == 1:
    print("CORRECT")
else:
    print("WRONG")
    print("Please try again.")

### 예제 4

다음 코드의 실행 결과로 알맞은 것을 적으시오.
```python
s = "Python is a high-level programming language."
print(s.split()[-2])
```

In [ ]:
정답 = "" #@param {type:"string"}
if 정답 == "programming":
    print("CORRECT")
else:
    print("WRONG")
    print("Please try again.")